src/SprykerEco/Zed/Braintree/Business/Payment/Transaction/Handler/RefundItemsTransactionHandler.php

```diff
@@ -0,0 +1,218 @@
+     */
+    protected function getRefund(array $salesOrderItems, SpySalesOrder $salesOrderEntity): RefundTransfer
+    {
+        return $this->refundFacade->calculateRefund($salesOrderItems, $salesOrderEntity);
+    }
+
+    /**
+     * @param \Orm\Zed\Sales\Persistence\SpySalesOrderItem[] $orderItems
+     *
+     * @return \ArrayObject
+     */
+    protected function getItemsForRefunding(array $orderItems): ArrayObject
+    {
+        $itemsForRefunding = [];
+
+        foreach ($orderItems as $orderItem) {
+            $itemTransfer = new ItemTransfer();
+            $itemTransfer->fromArray($orderItem->toArray(), true);
+            $itemsForRefunding[] = $itemTransfer;
+        }
+
+        return $this->getRefundedUniqItems($itemsForRefunding);
+    }
+
+    /**
+     * @param \Orm\Zed\Sales\Persistence\SpySalesOrderItem[] $salesOrderItems
+     * @param \Orm\Zed\Sales\Persistence\SpySalesOrder $salesOrderEntity
+     * @param \Generated\Shared\Transfer\RefundTransfer $refundTransfer
+     *
+     * @return \Generated\Shared\Transfer\TransactionMetaTransfer
+     */
+    protected function getTransactionMetaTransfer(array $salesOrderItems, SpySalesOrder $salesOrderEntity, RefundTransfer $refundTransfer): TransactionMetaTransfer
+    {
+        $transactionMetaTransfer = new TransactionMetaTransfer();
+        $transactionMetaTransfer->setIdSalesOrder($salesOrderEntity->getIdSalesOrder());
+        $transactionMetaTransfer->setRefund($refundTransfer);
+        $transactionMetaTransfer->setItems($this->getItemsForRefunding($salesOrderItems));
+
+        return $transactionMetaTransfer;
+    }
+
+    /**
+     * @param \ArrayObject|\Generated\Shared\Transfer\ItemTransfer[] $itemTransfers
+     *
+     * @return \ArrayObject|\Generated\Shared\Transfer\ItemTransfer[]
+     */
+    protected function getRefundedUniqItems(array $itemTransfers): iterable
+    {
+        $uniqItems = [];
+        $uniqItemIds = [];
+
+        foreach ($itemTransfers as $itemTransfer) {
+            if (!in_array($itemTransfer->getIdSalesOrderItem(), $uniqItemIds)) {
+                $uniqItems[] = $itemTransfer;
+                $uniqItemIds[] = $uniqItemIds;
+            }
+        }
+
+        $uniqItems = new ArrayObject($uniqItems);
+
+        return $uniqItems;
+    }
+
+    /**
+     * @param \Generated\Shared\Transfer\TransactionMetaTransfer $transactionMetaTransfer
+     *
+     * @return array
+     */
+    protected function getOrderItemsGroupedByTransaction(TransactionMetaTransfer $transactionMetaTransfer): array
+    {
+        $itemsByTransactions = [];
+
+        foreach ($transactionMetaTransfer->getItems() as $itemTransfer) {
+            $paymentBraintreeOrderItemTransfer = $this->braintreeRepository
+                ->findPaymentBraintreeOrderItemByIdSalesOrderItem($itemTransfer->getIdSalesOrderItem());
+
+            $paymentBraintreeTransactionSatusLogTransfer = $this->braintreeRepository
```

Comment:
Is it possible to make this queries in bulk? It's not a very good idea to run them in loop. We should try to avoid this is if we can.